perception/autoware_ground_segmentation/src/scan_ground_filter/node.cpp

```diff
@@ -336,239 +373,271 @@ void ScanGroundFilterComponent::classifyPointCloudGridScan(
-        checkContinuousGndGrid(*p, p_orig_point, gnd_grids);
-      } else if (
-        p->radius - gnd_grids.back().radius < gnd_grid_continual_thresh_ * curr_grid_size) {
-        checkDiscontinuousGndGrid(*p, p_orig_point, gnd_grids);
-      } else {
-        checkBreakGndGrid(*p, p_orig_point, gnd_grids);
-      }
-      if (p->point_state == PointLabel::NON_GROUND) {
-        out_no_ground_indices.indices.push_back(p->orig_index);
-      } else if (p->point_state == PointLabel::GROUND) {
-        ground_cluster.addPoint(p->radius, p_orig_point.z, p->orig_index);
       }
     }
-  }
+  }  // scope for time keeper ends
 }
 
 void ScanGroundFilterComponent::classifyPointCloud(
   const PointCloud2ConstPtr & in_cloud, std::vector<PointCloudVector> & in_radial_ordered_clouds,
   pcl::PointIndices & out_no_ground_indices)
 {
+  std::unique_ptr<ScopedTimeTrack> st_ptr;
+  if (time_keeper_) st_ptr = std::make_unique<ScopedTimeTrack>(__func__, *time_keeper_);
+
   out_no_ground_indices.indices.clear();
 
   const pcl::PointXYZ init_ground_point(0, 0, 0);
   pcl::PointXYZ virtual_ground_point(0, 0, 0);
   calcVirtualGroundOrigin(virtual_ground_point);
 
-  // point classification algorithm
-  // sweep through each radial division
-  for (size_t i = 0; i < in_radial_ordered_clouds.size(); ++i) {
-    float prev_gnd_radius = 0.0f;
-    float prev_gnd_slope = 0.0f;
-    PointsCentroid ground_cluster, non_ground_cluster;
-    PointLabel prev_point_label = PointLabel::INIT;
-    pcl::PointXYZ prev_gnd_point(0, 0, 0), p_orig_point, prev_p_orig_point;
-    // loop through each point in the radial div
-    for (size_t j = 0; j < in_radial_ordered_clouds[i].size(); ++j) {
-      float points_distance = 0.0f;
-      const float local_slope_max_angle = local_slope_max_angle_rad_;
-      prev_p_orig_point = p_orig_point;
-      auto * p = &in_radial_ordered_clouds[i][j];
-      get_point_from_global_offset(in_cloud, p_orig_point, in_cloud->point_step * p->orig_index);
-      if (j == 0) {
-        bool is_front_side = (p_orig_point.x > virtual_ground_point.x);
-        if (use_virtual_ground_point_ && is_front_side) {
-          prev_gnd_point = virtual_ground_point;
+  {  // add scope for time keeper
+    std::unique_ptr<ScopedTimeTrack> inner_st_ptr;
+    if (time_keeper_)
+      inner_st_ptr = std::make_unique<ScopedTimeTrack>("classify_loop", *time_keeper_);
+
+    // point classification algorithm
+    // sweep through each radial division
+    for (size_t i = 0; i < in_radial_ordered_clouds.size(); ++i) {
+      float prev_gnd_radius = 0.0f;
+      float prev_gnd_slope = 0.0f;
+      PointsCentroid ground_cluster, non_ground_cluster;
+      PointLabel prev_point_label = PointLabel::INIT;
+      pcl::PointXYZ prev_gnd_point(0, 0, 0), p_orig_point, prev_p_orig_point;
+
+      std::unique_ptr<ScopedTimeTrack> radial_loop_st_ptr;
+      if (time_keeper_)
+        radial_loop_st_ptr = std::make_unique<ScopedTimeTrack>("radial_loop", *time_keeper_);
```

Comment:
This time keeper is called too much.
The for loop is the only loop in the function `classifyPointCloud` and separating this is not so meaningful.

Reply:
As same as other part, I removed the timekeeper except the top level in https://github.com/autowarefoundation/autoware.universe/commit/45de8e7f4fac943213fc0a6fd54bdf45609b40ea .